server/sched.go:  Added support for grouping GPUs ( Power-Consumption + Low-PCIe Speed Performance gains)

dan-and opened the pull request:
This patch changes Ollama to allow grouping GPUs to memory-fit to the requested model, instead of the former algorithm of using one GPU <OR> distributing over all available GPUs. 

### Benefits:
* Lower amount of (PCIe-)bus communication between GPUs - especially when they are not very high speed  
* Allowing unallocated GPUs to get into power-saving mode. 

### How to use: 
* OLLAMA_SCHED_SPREAD=0 (or keep it unset) to use only one GPU, and then evenly distribute the load across all GPUs. (old default behavior) 
* OLLAMA_SCHED_SPREAD=1 (or set to anything else) to evenly distribute the load across all GPUs. (old default behavior)
* OLLAMA_SCHED_SPREAD=2 to keep the number of GPUs to a minimum and only use as many as needed to run the model. 

### Tests and Environment: 
Like so many in the home-lab community, my Ollama server is built on a budget, which resulted in a low-power optimized server with 7 GPUs (Nvidia RTX 3060 - 12GB) with a total of 72GB VRAM. 

When using models which require around 30-40 GB, all GPUs have to run, which results in a higher power consumption. With this patch, just a few models are in use, which powers down all other GPUs to around 3 watts. 
Since the server (Ryzen 5500G / Asus Prime B450)  is not enterprise grade, the PCIe bus far slower, which also slows down the utilization when using too many GPUs with one single model. 

### Results (Power consumption measured "at the wall") 

Qwen3:14b - 7d7da67570e2 - OLLAMA_SCHED_SPREAD unset:  448 Watt with 27 fps  
Qwen3:14b - 7d7da67570e2 - OLLAMA_SCHED_SPREAD=1:  448 Watt with 27 fps  
Qwen3:14b - 7d7da67570e2 - OLLAMA_SCHED_SPREAD=2:  335 Watt with 29 fps  

Qwen3:30b - 2ee832bc15b5 - OLLAMA_SCHED_SPREAD unset:  391 Watt with 41.1 fps  
Qwen3:30b - 2ee832bc15b5 - OLLAMA_SCHED_SPREAD=1:  391 Watt with 41.1 fps  
Qwen3:30b - 2ee832bc15b5 - OLLAMA_SCHED_SPREAD=2:  311 Watt with 46.7 fps  

(Yes, Qwen3:14b is far more inefficient on my system than the slightly younger 30b variant) 

gemma3:4b - f4031aab637d - OLLAMA_SCHED_SPREAD unset: 270 Watt with 69.4 fps 
gemma3:4b - f4031aab637d - OLLAMA_SCHED_SPREAD=1 : 389 Watt with 54.1 fps 
gemma3:4b - f4031aab637d - OLLAMA_SCHED_SPREAD=2 : 300 Watt with 60.2 fps 

(gemma3:4b with num_ctx of 8192 resulted in utilizing 2 GPUs with OLLAMA_SCHED_SPREAD=2 whereas one GPU would have been enough, as the estimation is more conservative)

igs-raiderler commented:
Interesting solution to this problem! I've been working on similar challenges recently.

dan-and commented:
Is there anything I can do to get it reviewed?

dan-and commented:
@rick-github , thanks for pointing me to the VRAM allocation effect of my patch, as this is next to the power-consumption is a huge benefit. 

There are examples of how much VRAM overhead ollama is consuming, when we are using the tactic of spreading the model over all available GPUs instead of grouping them.

dan-and commented:
Hi @jmorganca and @mxyng 

Do you see major issues in my patch, which may prevent including it into the master branch? If so, please give me feedback so that I can update accordingly 

Kind regards

dan-and commented:
Hi Jesse,
thanks for the feedback. 
I will rearrange my patch so that it is the default behavior. Give me 2-3 days and it will be updated.

dan-and commented:
@jessegross  
Updated to make it the default behavior, and OLLAMA_SCHED_SPREAD=1 will spread it evenly.

Because of my patch, I have introduced two slog.Info and one slog.Debug into the code. It helped me quite well to figure out if ollama is doing the correct allocation with various system setups and configs. Should I keep them in the code, or would you suggest  removing them or moving the slog.Info to slog.Debug to keep the amount of logs low? 

Regards,
Daniel 

PS: It works great on my 7 GPU system:  

OLLAMA_SCHED_SPREAD=1  

> daniel@gpu:~/source/ollama_dan$ ollama run qwen3:4b --verbose "say hi!" 
> Thinking...
> Okay, the user said "say hi!" so I need to respond with a friendly greeting. Let me make sure to keep it warm and welcoming. Maybe add an emoji to make it more lively. I should also ask how I can assist them today. Let me 
> check the tone—should be positive and eager to help. Alright, that should do it.
> ...done thinking.
> 
> Hello! 😊 How can I assist you today?
> 
> total duration:       11.976054545s
> load duration:        **10.255610156s**
> prompt eval count:    13 token(s)
> prompt eval duration: 358.652306ms
> prompt eval rate:     36.25 tokens/s
> eval count:           88 token(s)
> eval duration:        1.36080978s
> eval rate:            **64.67 tokens/s**
> daniel@gpu:~/source/ollama_dan$ ollama ps
> NAME        ID              SIZE     PROCESSOR    CONTEXT    UNTIL               
> qwen3:4b    2bfd38a7daaf    **20 GB**    100% GPU     8192       29 minutes from now    
> 


OLLAMA_SCHED_SPREAD=0 or not set at all

> daniel@gpu:~/source/ollama_dan$ ollama run qwen3:4b --verbose "say hi!"
> Thinking...
> Okay, the user said "say hi!" so I need to respond in a friendly and welcoming manner. Let me start by acknowledging their greeting. Maybe use an emoji to keep it light and approachable.
> I should mention my name, Qwen, to establish identity. Then, offer assistance by asking how I can help them today. Keep it concise but open-ended so they feel comfortable to share their
> needs. Make sure the tone is positive and eager to help. Let me put that all together.
> ...done thinking.
> 
> Hello! I'm Qwen, your helpful assistant. How can I assist you today? 😊
> 
> total duration:       5.474187439s
> load duration:        **3.6653167s**
> prompt eval count:    13 token(s)
> prompt eval duration: 139.89265ms
> prompt eval rate:     92.93 tokens/s
> eval count:           125 token(s)
> eval duration:        1.668211081s
> eval rate:            **74.93 tokens/s**
> daniel@gpu:~/source/ollama_dan$ ollama ps
> NAME        ID              SIZE      PROCESSOR    CONTEXT    UNTIL
> qwen3:4b    2bfd38a7daaf    **7.3 GB**    100% GPU     8192       29 minutes from now

jessegross commented:
Yes, I think you can remove the extra logging and the memory calculations that go into them. Please also retain existing comments and TODOs that aren't affected by your change.

Since you are iterating from smallest to largest number of GPUs, once you find a set that fits, you can just return that. You don't need to keep looking and find the best one - no later tries are going to use fewer GPUs.

I have some in-progress code that does something similar for a different code path, you can use it as a reference:
https://github.com/ollama/ollama/blob/f4c2e622425e4f987f664e16f60a62b0a4c4e070/llm/server.go#L777

dan-and commented:
@jessegross  any additional comments and hints for me?
